stdlib/src/builtin/string.mojo

```diff
@@ -660,6 +666,87 @@ fn isprintable(c: UInt8) -> Bool:
 # ===----------------------------------------------------------------------=== #
 
 
+fn _utf8_byte_type(b: UInt8) -> UInt8:
+    if b < 0b1000_0000:
+        return 0
+    elif b < 0b1100_0000:
+        return 1  # is continuation byte
+    elif b < 0b1110_0000:
+        return 2  # is 2 byte long
+    elif b < 0b1111_0000:
+        return 3  # is 3 byte long
+    else:
+        return 4  # is 4 byte long
+
+
+@value
+struct _StringIter[
+    is_mutable: Bool,
+    lifetime: AnyLifetime[is_mutable].type,
+    forward: Bool = True,
+]:
+    """Iterator for String.
+
+    Parameters:
+        is_mutable: Whether the slice is mutable.
+        lifetime: The lifetime of the underlying string data.
+        forward: The iteration direction. `False` is backwards.
+    """
+
+    var index: Int
+    var continuation_bytes: Int
+    var ptr: UnsafePointer[UInt8]
+    var len: Int
+
+    fn __init__(inout self, unsafe_pointer: UnsafePointer[UInt8], len: Int):
```

Comment:
nit: to avoid some nasty errors when someone tries to use `len(...)`
```suggestion
    fn __init__(inout self, unsafe_pointer: UnsafePointer[UInt8], length: Int):
```